pkg/cmd/ensureenv/ensureenv.go

```diff
@@ -0,0 +1,137 @@
+	}
+
+	ensure.AddFlags(cmd.Flags())
+	return cmd
+}
+
+func (e *ensureOpts) AddFlags(fs *pflag.FlagSet) {
+	fs.BoolVar(&e.allowInvalidNodeIP, "allow-invalid-node-ip", true, "If this is set and the value of --node-ip is invalid (not equal to environment variables given by --ip-ev-equals), continue validation and check the value of --current-revision-node-ip against the network interfaces.")
+	fs.StringVar(&e.nodeIP, "node-ip", "", "The internal IP of the node provided by the downward API.")
+	fs.StringVar(&e.currentRevNodeIP, "current-revision-node-ip", "", "If --node-ip is invalid and --allow-invalid-node-ip is true, check the network interfaces for an IP address matching this value.")
+	fs.StringSliceVar(&e.notEmpty, "check-set-and-not-empty", []string{}, "Validate that the given environment variable(s) are set and have non-empty values.")
+}
+
+func (e *ensureOpts) Validate() (err error) {
+	// Validate that the incoming environment variables are set and have non-empty values.
+	for _, ev := range e.notEmpty {
+		if value, has := os.LookupEnv(ev); !has {
+			return fmt.Errorf("the value of environment variable %v must be set.", ev)
+		} else if value == "" {
+			return fmt.Errorf("the value of environment variable %v must not be empty.", ev)
+		}
+	}
+
+	// Check node-ip must be set, unless we accept invalid (not set or empty) values.
+	if e.nodeIP == "" && !e.allowInvalidNodeIP {
+		return fmt.Errorf("since --allow-invalid-node-ip is not set, --node-ip must be set to a non-empty value.")
+	}
+
+	// Current revision must also be set.
+	if e.currentRevNodeIP == "" {
+		return fmt.Errorf("--current-revision-node-ip must be set to a non-empty value.")
+	}
+
+	return nil
+}
+
+func (e *ensureOpts) Run() (err error) {
+	// If the node-ip isn't valid, we'll check the network interfaces.
+	if e.nodeIP != e.currentRevNodeIP {
+		if !e.allowInvalidNodeIP {
+			return fmt.Errorf("since --allow-invalid-node-ip is not set, node-ip and current-revision-node-ip must be equal.")
+		}
+
+		// Since the value of ip-ev is empty, we're going to check our
+		//   fallback environment variable against our network
+		//   interfaces, to see if we find a match.
+		// If we find a match, we'll consider ourselves valid, if we can't
+		//   we'll fail.
+		var interfaces []net.Interface
+		if interfaces, err = net.Interfaces(); err != nil {
+			return err
+		}
+
+		// Parse the current-revision-node-ip value into an IP structure to test for equality later.
+		var currentRevIP = net.ParseIP(e.currentRevNodeIP)
+		for _, iface := range interfaces {
+			var addrs []net.Addr
+
+			if addrs, err = iface.Addrs(); err != nil {
+				return err
+			}
+
+			for _, addr := range addrs {
+				switch ip := addr.(type) {
+				// Casting here to remove the mask before checking for equality.
+				case *net.IPNet:
+					if ip.IP.Equal(currentRevIP) {
+						// We found a match, we'll assume now that the value of fallbackEV is good
+						//		and we'll continue allowing etcd to start.
```

Comment:
Nit:
```suggestion
						// and we'll continue allowing etcd to start.
```

Reply:
But you can do this in a follow up if you want since the PR is mostly good to go.